### PR TITLE
Update README with documentation on setting headers

### PR DESCRIPTION
**Pull request by nmiyake**
Document setting headers at api level using @Headers
or at client level using RequestInterceptor or Target.

### CODE REVIEW

**Comment by nmiyake** [comment]
@adriancole

File: README.md [hunk 1]
```diff
@@ -246,6 +246,70 @@ client.xml("denominator", "secret"); // <login "user_name"="denominator" "passwo
...
+
+```java
+ @RequestLine("POST /")
+ @Headers({
+  "X-Foo: Bar",
+  "X-Ping: {token}"
+ }) void post(@Param("token") String token);
+```
+
+These approaches specify specific header entries as part of the api without requiring any cutomizations
```
Comment (adriancole) [inline README.md]: cutomizations

File: README.md [hunk 2]
```diff
@@ -246,6 +246,70 @@ client.xml("denominator", "secret"); // <login "user_name"="denominator" "passwo
...
+when buildling Feing clients. It is not currently possible to customize the header entries themselves
+on a per-request basis at the api level.
+
+#### Set headers using client
```
Comment (adriancole) [inline README.md]: Setting headers per target

File: README.md [hunk 3]
```diff
@@ -246,6 +246,70 @@ client.xml("denominator", "secret"); // <login "user_name"="denominator" "passwo
...
+
+```java
+  static class ForwardedForTarget extends HardCodedTarget {
+    public ForwardedForTarget(Class type, String url) {
+      super(type, url);
+    }
+
+    @Override
+    public Request apply(RequestTemplate input) {
+      input.header("X-Forwarded-For", "origin.host.com");
```
Comment (adriancole) [inline README.md]: not a very realistic target..

Target's main use has been adding authentication headers or anything tha varies on the url.
Have a look at denominator for examples. https://github.com/Netflix/denominator/find/master

File: README.md [hunk 4]
```diff
@@ -246,6 +246,70 @@ client.xml("denominator", "secret"); // <login "user_name"="denominator" "passwo
...
+      return super.apply(input);
+    }
+  }
+  ...
+  Bank bank = Feign.builder()
+          .decoder(accountDecoder)
+          .target(new ForwardedForTarget(Bank.class, "https://api.examplebank.com"));
+```
+
+These approaches depend on the custom `RequestInterceptor` or `Target` being set on the Feign
```
Comment (adriancole) [inline README.md]: Maybe reword it, as many things are implicit, and this is not different than things like encoders, or anything else in feign builder. What's important is the calling thread.

When setting dynamic headers with  `RequestInterceptor` or `Target`, bear in mind that `apply` is called as a side effect of an api call: that can be made from any thread.

In simplest case, all callers get the same value. For example, an authentication token is set in a thread-safe fashion and shared for all callers.

Another use case is context-specific values. For example, an tracing id is set in a thread-local upstream to Feign. The target or request interceptor reads that context and adds a trace header accordingly.

**Comment by adriancole** [comment]
looks good. main recommendation is to use concrete examples here. I think authentication token and trace ids explain the thread-safety concern very well. Plus, in most cases, you'd never want the user to have to explicitly provide a auth token or trace context.. this reinforces common implicit uses and the underlying point about thread safety.

hope this helps

**Comment by nmiyake** [comment]
Thanks for the feedback -- made the example more realistic (setting an auth token and thread-local trace id) and provided more specific examples in documentation.

**Comment by adriancole** [comment]
Nicely done.
